Add NumericWithUnit and SmartText Widgets

mockoocy opened the pull request:
Closes #1897
This provides a way to display unit for numeric fields. By default RJSF renders the numeric fields with a built-in `TextWidget`. Thus it would be required for user to always specify in the `uiSchema` widget to be new `NumericWithUnitWidget`.
That wouldn't be convenient, so the `TextWidget` is being overriden with `SmartTextWidget`, which just renders `NumericWithUnitWidget` if the field is numeric.


This is how it looks
<img width="351" height="84" alt="image" src="https://github.com/user-attachments/assets/1f22af7e-60f0-420e-a6e6-a500ce1ce63b" />

It requires setting a `unit` field in the pydantic model, like so:
```python
class TestUserCollectionParameters(BaseModel):
    num_images: int = Field(0, description="")
    exp_time: float = Field(100e-6, gt=0, lt=1, unit="s", description="s")
```

Once it were to be merged, I'd add documentation section for it, along making documentation changes described in the #1894.

Note: feel free to propose alternative names for these widgets, the current names are not really that good... :D

mockoocy commented:
Note: It is a different approach from the one described in the linked issue. It comes from some "experimental" branch of our fork, as we were also trying to make some more complex forms with it :)
This one seems nicer anyway ;)

axelboc commented on ui/src/components/Tasks/Widgets/NumericWithUnitWidget/NumericWithUnitWidget.jsx:
```diff
@@ -0,0 +1,42 @@
+      <Form.Control
+        id={id}
+        type="number"
+        className={styles.input} /* ← gives padding-right */
```
```suggestion
        className={styles.input} // gives padding-right
```

axelboc commented on ui/src/components/Tasks/Widgets/NumericWithUnitWidget/NumericWithUnitWidget.module.css:
```diff
@@ -0,0 +1,20 @@
+  white-space: nowrap;
+  font-size: 0.875rem;
+  color: var(--bs-secondary);
+  pointer-events: none; /* don’t block clicks on the input */
```
Best to avoid special characters :wink: 

```suggestion
  pointer-events: none; /* don't block clicks on the input */
```

mockoocy replied:
Sorry for these. I did not take a good enough look at these

axelboc commented on ui/src/components/Tasks/Widgets/NumericWithUnitWidget/NumericWithUnitWidget.jsx:
```diff
@@ -0,0 +1,42 @@
+ * @param {import('@rjsf/utils').WidgetProps} props
+ * @returns
+ */
+export default function NumericWithUnitWidget({
```
This is used for all numeric inputs, not just those with units, so I would rename it `NumericWidget`

axelboc commented on ui/src/components/Tasks/GenericTaskForm.jsx:
```diff
@@ -313,6 +314,9 @@ class GenericTaskForm extends React.Component {
                 FieldErrorTemplate: CustomFieldErrorTemplate,
               }}
+              widgets={{
+                TextWidget: SmartTextWidget,
```
Maybe `TextOrNumericWidget`?

marcus-oscarsson replied:
yes, that reads nicer

axelboc reviewed (approved):
Looks great!

marcus-oscarsson commented:
Well done @mockoocy and big thumps up for the reactivity super fast !

marcus-oscarsson commented:
I have one minor comment, I also think it already looks great but I think the unit is easy to miss, would it help to make the text a bit bigger or bold (or maybe both) ?

marcus-oscarsson commented:
I like the use of "unit" in the pydantic model very good !

mockoocy commented:
> I have one minor comment, I also think it already looks great but I think the unit is easy to miss, would it help to make the text a bit bigger or bold (or maybe both) ?

I made the text a bit bigger (uses the same font size as the input text). 
I see that there is a problem for wide inputs, where the unit is far away from the text. For now I've decided to use some lighter colour to make it standout more. 
<img width="361" height="69" alt="image" src="https://github.com/user-attachments/assets/379052e1-241d-4b09-b190-a19a6d31f2e6" />

I also see two ways it could be improved
1. there could be a different background around the unit
2. the unit could be displayed right after the input text

But for now this looks alright IMO

marcus-oscarsson commented:
Great for the font size, I would probably no use cyan for the color. I think black or very dark gray is fine. You could maybe try to add it after the label with parentheses or brackets around it. i.e **Exp. Time _[s]_** or **Exp. Time [s]**

walesch-yan commented:
I agree with Marcus on the unit being easy to miss and I also do not like the cyan color, I used a plugin for colorblindness and especially when having a weakness to the color green, the cyan color becomes a very light gray, difficult to see in that case in my opinion. Otherwise, great work on the PR!

mockoocy committed:
Custom forms: Add unit next to field labels

If unit for a field is specified in the pydantic model, it will be
rendered next to it's label.
The unit can be specified like so, in the QueueEntry definition:
```python
class TestUserCollectionParameters(BaseModel):
    num_images: int = Field(0, description="")
    exp_time: float = Field(100e-6, gt=0, lt=1, unit="s", description="s")
```

mockoocy commented:
Okay, I went with the unit in the square brackets.
<img width="372" height="87" alt="image" src="https://github.com/user-attachments/assets/f3780e66-8b8f-4612-be26-5d9a53270a2c" />
So there will be less code added, so that's also nice I suppose :D

marcus-oscarsson commented:
Excellent well done !

marcus-oscarsson commented:
Thanks !
